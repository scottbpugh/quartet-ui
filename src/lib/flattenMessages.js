// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

export function flattenMessages(nestedMessages, prefix = "") {
  return Object.keys(nestedMessages).reduce((messages, key) => {
    const value = nestedMessages[key];
    const prefixedKey = prefix ? `${prefix}.${key}` : key;

    if (typeof value === "string") {
      messages[prefixedKey] = value;
    } else {
      Object.assign(messages, flattenMessages(value, prefixedKey));
    }

    return messages;
  }, {});
}

window.qu4rtet.exports("lib/flattenMessages", this);
