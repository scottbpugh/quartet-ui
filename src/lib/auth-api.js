// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import {showMessage} from "lib/message";
import {getSyncValidators} from "components/elements/forms";
import {pluginRegistry} from "plugins/pluginRegistration";

/**
 * prepHeaders - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers.
 */
export const prepHeaders = (server, method = "GET") => {
  let headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  return {
    method: method,
    headers: headers,
    mode: "cors"
  };
};

/**
 * prepHeaders - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers and basic auth.
 */
export const prepHeadersAuth = async (server, method = "GET") => {
  let headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  headers.append(
    "Authorization",
    await pluginRegistry.getServer(server.serverID).getAuthorization()
  );
  return {
    method: method,
    headers: headers,
    mode: "cors"
  };
};

/**
 * getFormInfo - Description
 *
 * @param {A server object} server       Description
 * @param {Path to endpoint} path         Description
 * @param {Callback to set the state with formStructure} createForm   Description
 * @param {type} processField Description
 *
 */
export const getFormInfo = async (server, path, createForm, processField) => {
  return fetch(`${server.url}${path}`, await prepHeadersAuth(server, "OPTIONS"))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      // parse the values and filter to the one that are not readonly.
      let postFields = data.actions.POST;
      let formStructure = Object.keys(postFields)
        .map(field => {
          if (postFields[field].read_only === false) {
            return {name: field, description: postFields[field]};
          } else {
            return null;
          }
        })
        .filter(fieldObj => {
          if (processField) {
            return processField(fieldObj);
          }
          if (fieldObj) {
            // create sync validation arrays.
            fieldObj.validate = getSyncValidators(fieldObj);
            return true;
          }
          return false;
        });
      createForm(formStructure);
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "app.servers.errorFormFetch",
        values: {error: error, serverName: server.serverSettingName}
      });
      throw error;
    });
};
