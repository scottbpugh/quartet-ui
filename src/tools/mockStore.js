// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React from "react";

import {IntlProvider} from "react-intl";
import {addLocaleData} from "react-intl";
import messages from "messages";
import en from "react-intl/locale-data/en";
import fr from "react-intl/locale-data/fr";
import {flattenMessages} from "lib/flattenMessages";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

addLocaleData([...en, ...fr]);
let defaultLocale = "en-US";
const middlewares = [thunk];

export const initialState = {
  dashboard: {notifications: []},
  serversettings: {
    servers: {
      "824ae2dd-2411-4adf-a846-d9b310a33d4b": {
        serverSettingName: "Server #1",
        serverName: "localhost",
        port: "80",
        ssl: false,
        username: "root",
        password: "toor",
        serverID: "824ae2dd-2411-4adf-a846-d9b310a33d4b"
      }
    }
  },
  intl: {
    defaultLocale: defaultLocale,
    locale: defaultLocale,
    messages: flattenMessages(messages[defaultLocale])
  },
  layout: {pageTitle: {id: "app.nav.servers"}},
  plugins: {plugins: {}, navTreeItems: []}
};

export const mockStore = configureStore(middlewares);

export const TestWrapper = ({locale, messages, children}) => {
  if (locale === undefined) {
    locale = defaultLocale;
  }
  if (messages === undefined) {
    messages = initialState.intl.messages;
  }
  return (
    <IntlProvider locale={locale} messages={messages}>
      {children}
    </IntlProvider>
  );
};

