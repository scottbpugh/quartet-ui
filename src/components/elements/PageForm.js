// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React, {Component} from "react";
import {showMessage} from "lib/message";
import {DefaultField} from "components/elements/forms";
import {Field, SubmissionError} from "redux-form";
import {getFormInfo} from "lib/server-api";
import {connect} from "react-redux";
import {withRouter} from "react-router";
import {Callout, Intent, FormGroup} from "@blueprintjs/core";

class _PageForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      formStructure: []
    };
    this.formObject = {}; // populated for updates.
    this.formStructureRetrieved = false;
  }
  componentDidMount() {
    this.constructForm(this.props);
  }
  submit = postValues => {
    let {
      server,
      operationId,
      prepopulatedValues,
      objectName,
      redirectPath,
      parameters,
      submitCallback
    } = this.props;
    if (prepopulatedValues) {
      for (let field of prepopulatedValues) {
        // replaces/sets programmatically.
        postValues[field.name] = field.value;
      }
    }
    let validatedData = {};
    // set to null if empty string
    Object.keys(postValues).forEach(item => {
      validatedData[item] = postValues[item] === "" ? null : postValues[item];
    });
    if (parameters) {
      parameters.data = validatedData;
    } else {
      parameters = {data: validatedData};
    }
    return server.getClient().then(client => {
      return client
        .execute({
          operationId: operationId,
          parameters: parameters,
          securities: {
            authorized: client.securities,
            specSecurity: [client.spec.securityDefinitions]
          }
        })
        .then(result => {
          if (submitCallback) {
            // execute post submit logic...
            submitCallback();
          }
          if (result.status === 201) {
            showMessage({
              msg: `New ${objectName} created successfully`,
              type: "success"
            });
          } else if (result.status === 200) {
            showMessage({
              msg: `Existing ${objectName} updated successfully`,
              type: "success"
            });
          }
          if (redirectPath) {
            this.props.history.push(redirectPath);
          }
        })
        .catch(error => {
          if (error.status === 400 && error.response && error.response.body) {
            if ("non_field_errors" in error.response.body) {
              // a form-wide error is present.
              throw new SubmissionError({
                ...error.response.body,
                _error: error.response.body.non_field_errors
              });
            }
            // we have an object with validation errors.
            throw new SubmissionError(error.response.body);
          }
          if (error.message) {
            showMessage({
              id: "app.common.mainError",
              values: {msg: error.message},
              type: "error"
            });
          }
        });
    });
  };

  constructForm = props => {
    const {djangoPath} = props;
    if (!this.formStructureRetrieved) {
      let createForm = formStructure => {
        this.setState(
          {
            formStructure: formStructure
          },
          () => {
            if (props.existingValues) {
              props.initialize(props.existingValues);
            }
          }
        );
        this.formStructureRetrieved = true;
      };
      getFormInfo(props.server, djangoPath, createForm);
    }
  };
  render() {
    const {
      error,
      handleSubmit,
      submitting,
      prepopulatedValues,
      fieldElements
    } = this.props;
    let form = this.state.formStructure
      .map(field => {
        let type = "text";
        if (field.description.type === "integer") {
          type = "number";
        } else if (field.description.type === "boolean") {
          type = "checkbox";
        } else if (
          field.description.type === "password" ||
          field.name.includes("password")
        ) {
          type = "password";
        }

        let filtered = [];

        if (prepopulatedValues) {
          filtered = prepopulatedValues.map(field => {
            return field.name;
          });
        }
        if (fieldElements && field.name in fieldElements) {
          // return the element as is with descriptions.
          return (
            <FormGroup
              helperText={field.description.help_text}
              label={field.description.label}
              required={field.description.required}>
              {fieldElements[field.name]}
            </FormGroup>
          );
        }
        if (!filtered.includes(field.name)) {
          if (field.description.type === "choice") {
            return (
              <Field
                key={field.name}
                name={field.name}
                type="select"
                component={DefaultField}
                width={300}
                fieldData={field}
                validate={field.validate}>
                <option value="" />
                {field.description.choices.map(choice => {
                  return (
                    <option value={choice.value}>{choice.display_name}</option>
                  );
                })}
              </Field>
            );
          }

          return (
            <Field
              key={field.name}
              name={field.name}
              fieldData={field}
              type={type}
              component={DefaultField}
              className="pt-input"
              width={300}
              validate={field.validate}
            />
          );
        } else {
          return null;
        }
      })
      .filter(field => {
        if (field) {
          return true;
        }
        return false;
      });
    return (
      <div>
        <form onSubmit={handleSubmit(this.submit.bind(this))}>
          {form}
          <button
            className="pt-button pt-intent-primary"
            type="submit"
            disabled={submitting}>
            Submit
          </button>
          {error ? (
            <Callout iconName="warning" intent={Intent.DANGER}>
              {error}
            </Callout>
          ) : null}
        </form>
      </div>
    );
  }
}

export default connect((state, ownProps) => {
  return {
    servers: state.serversettings.servers
  };
}, {})(withRouter(_PageForm));
