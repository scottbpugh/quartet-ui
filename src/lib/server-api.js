// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import {showMessage} from "lib/message";
import {getSyncValidators} from "components/elements/forms";
import {pluginRegistry} from "plugins/pluginRegistration";

/**
 * prepHeadersAuth - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers and basic auth.
 */
export const prepHeadersAuth = async (server, method = "GET") => {
  const headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  headers.append(
    "Authorization",
    await pluginRegistry.getServer(server.serverID).getAuthorization()
  );
  return {
    method,
    headers,
    mode: "cors"
  };
};

/**
 * getFormInfo - Returns an object with a form structure to be used by PageForm/redux form
 *
 * @param {A server object} server       Description
 * @param {Path to endpoint} path         Description
 * @param {Callback to set the state with formStructure} createForm   Description
 * @param {type} processField Description
 *
 */
export const getFormInfo = async (server, path, createForm, processField) => {
  return fetch(`${server.url}${path}`, await prepHeadersAuth(server, "OPTIONS"))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      // parse the values and filter to the one that are not readonly.
      try {
        const postFields = data.actions.POST;
        const formStructure = Object.keys(postFields)
          .map(field => {
            if (
              field === "id" ||
              (postFields[field].type === "field" &&
                postFields[field].read_only === true)
            ) {
              return null;
            }
            return {name: field, description: postFields[field]};
          })
          .filter(fieldObj => {
            if (processField) {
              return processField(fieldObj);
            }
            if (fieldObj) {
              // create sync validation arrays.
              fieldObj.validate = getSyncValidators(fieldObj);
              return true;
            }
            return false;
          });
        createForm(formStructure);
      } catch (e) {
        if (e.message.startsWith("Cannot read property 'POST'")) {
          pluginRegistry.getHistory().push("/access-denied");
        }
      }
    })
    .catch(error => {
      debugger;
      console.log("About to redirect to access denied");
      //pluginRegistry.getHistory().push("/access-denied");
      showMessage({
        type: "danger",
        id: "app.servers.errorFormFetch",
        values: {error, serverName: server.serverSettingName}
      });
      throw error;
    });
};

/**
 * fetchObject - fetches a single object from any Swagger API endpoint.
 *
 * @param {Server}   serverInstance  A Server instance
 * @param {string} [operationId=]  The Swagger operation id
 * @param {object} [parameters={}] Parameter passed with the call.
 *
 * @return {object} An object from the API call.
 */
export const fetchObject = async (
  serverInstance,
  operationId = "",
  parameters = {}
) => {
  const client = await serverInstance.getClient();
  try {
    const response = await client.execute({
      operationId,
      parameters,
      securities: {
        authorized: client.securities,
        specSecurity: [client.spec.securityDefinitions]
      }
    });
    if (response.ok) {
      return response.body;
    }
  } catch (e) {
    showMessage({
      type: "error",
      id: "plugins.numberRange.errorVanilla",
      values: {error: e}
    });
    throw e;
  }
};

/**
 * fetchPageList - Fetches a single list page.
 *
 * @param {type} serverInstance A Server instance
 * @param {type} operationId    The Swagger
 * @param {type} parameters     Parameters passed to the call (page...)
 *
 * @return {object} A response body
 */
export const fetchPageList = async (
  serverInstance,
  operationId,
  parameters
) => {
  const client = await serverInstance.getClient();
  try {
    const response = await client.execute({
      operationId,
      parameters,
      securities: {
        authorized: client.securities,
        specSecurity: [client.spec.securityDefinitions]
      }
    });
    if (response.ok) {
      return response.body;
    }
    throw new Error(response);
  } catch (e) {
    showMessage({
      type: "error",
      id: "plugins.numberRange.errorVanilla",
      values: {error: e}
    });
    throw e;
  }
};

/**
 * fetchListAll - Loops through all pages of a list and concats the result.
 *
 * @param {Server}   serverInstance  A Server Instance
 * @param {string} [operationId=]  Swagger operation id
 * @param {object} [parameters={}] Parameters passed to the call.
 * @param {array}  [results=[]]    The results, all objects for this API list.
 *
 * @return {type} Description
 */
export const fetchListAll = async (
  serverInstance,
  operationId = "",
  parameters = {},
  results = []
) => {
  return new Promise((resolve, reject) => {
    serverInstance
      .getClient()
      .then(client => {
        client
          .execute({
            operationId,
            parameters,
            securities: {
              authorized: client.securities,
              specSecurity: [client.spec.securityDefinitions]
            }
          })
          .then(response => {
            if (response.ok) {
              if (Array.isArray(response.body)) {
                results = results.concat(response.body);
                resolve(results);
              } else if (response.body.results) {
                results = results.concat(response.body.results);
                // pagination in effect. Assuming page number navigation.
                if (response.body.next) {
                  const url = new URL(response.body.next);
                  const page = new URLSearchParams(url.search).get("page");
                  const subParameters = {...parameters, page};
                  fetchListAll(operationId, subParameters, results)
                    .then(resolve)
                    .catch(reject);
                } else {
                  resolve(results);
                }
              }
            } else {
              reject(response);
            }
          })
          .catch(e => {
            reject(e);
          });
      })
      .catch(e => {
        reject(e);
      });
  });
};

/**
 * deleteObject - Deletes an object.
 *
 * @param {string} [operationId=]  The operationId to execute.
 * @param {object} [parameters={}] Object data
 *
 * @return {object} response.
 */
export const deleteObject = async (
  serverInstance,
  operationId = "",
  parameters = {}
) => {
  const client = await serverInstance.getClient();
  const response = await client.execute({
    operationId,
    parameters,
    securities: {
      authorized: client.securities,
      specSecurity: [client.spec.securityDefinitions]
    }
  });
  if (response.ok) {
    return response;
  }
};

window.qu4rtet.exports("lib/server-api", this);
