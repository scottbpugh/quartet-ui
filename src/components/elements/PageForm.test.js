// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import React from "react";
import "tools/mockStore"; // mock ipcRenderer, localStorage, ...
import renderer from "react-test-renderer";
import PageForm from "./PageForm";
import {TestWrapper} from "tools/mockStore";
import Enzyme, {mount, shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import sinon from "sinon";
import toJson from "enzyme-to-json";
import * as authAPI from "lib/server-api";
import {reduxForm, change as changeFieldValue} from "redux-form";

// a real life example of a form options response to test form rendering.
// this is taking a shortcut and formatting the data the way getFormInfo does it.
const formattedFields = [
  {
    name: "GLN13",
    description: {
      type: "string",
      required: false,
      read_only: false,
      label: "GLN13",
      help_text:
        "The GLN (Global Location Number) provides a standard means to identify legal entities, trading parties and locations to support the requirements of electronic commerce. The GLN-13 is defined by GS1",
      max_length: 13
    },
    validate: null
  },
  {
    name: "SGLN",
    description: {
      type: "string",
      required: false,
      read_only: false,
      label: "SGLN",
      help_text:
        "The SGLN EPC scheme is used to assign a unique identity to a physical location or sub-location, such as a specific building or a specific unit of shelving within a warehouse.  TheSGLN is expressed as a URN value.",
      max_length: 150
    },
    validate: null
  },
  {
    name: "name",
    description: {
      type: "string",
      required: true,
      read_only: false,
      label: "Name",
      help_text: "A unique name for the location or party.",
      max_length: 128
    },
    validate: null
  },
  {
    name: "address1",
    description: {
      type: "string",
      required: false,
      read_only: false,
      label: "Street Address One",
      help_text:
        "For example, the name of the street and the number in the street or the name of a building",
      max_length: 1000
    },
    validate: null
  },
  {
    name: "address2",
    description: {
      type: "string",
      required: false,
      read_only: false,
      label: "Street Address Two",
      help_text:
        "The second free form line complements the first free form line to locate the party or location.",
      max_length: 1000
    },
    validate: null
  },
  {
    name: "address3",
    description: {
      type: "string",
      required: false,
      read_only: false,
      label: "Street Address Three",
      help_text:
        "The third free form line complements the first and second free form lines where necessary.",
      max_length: 1000
    },
    validate: null
  },
  {
    name: "country",
    description: {
      type: "string",
      required: false,
      read_only: false,
      label: "Country Code",
      help_text: "Country ISO 3166-1 alpha-2 Code",
      max_length: 2
    },
    validate: null
  },
  {
    name: "city",
    description: {
      type: "string",
      required: false,
      read_only: false,
      label: "City",
      help_text: "City",
      max_length: 50
    },
    validate: null
  },
  {
    name: "state_province",
    description: {
      type: "string",
      required: false,
      read_only: false,
      label: "State or Province",
      help_text:
        "One of the constituent units of a nation having a federal government.",
      max_length: 20
    },
    validate: null
  },
  {
    name: "postal_code",
    description: {
      type: "string",
      required: false,
      read_only: false,
      label: "Postal Code",
      help_text: "Postal Code",
      max_length: 20
    },
    validate: null
  },
  {
    name: "latitude",
    description: {
      type: "decimal",
      required: false,
      read_only: false,
      label: "Latitude",
      help_text:
        " Latitude of the location, in degrees. Positive numbers are northern latitude; negative numbers are southern latitude."
    },
    validate: null
  },
  {
    name: "longitude",
    description: {
      type: "decimal",
      required: false,
      read_only: false,
      label: "Longitude",
      help_text:
        "Longitude of the location, in degrees. Positive numbers are eastern longitude; negative numbers are western longitude."
    },
    validate: null
  },
  {
    name: "icon",
    description: {
      type: "file upload",
      required: false,
      read_only: false,
      label: "Icon",
      help_text: "An icon to represent the location in a GUI or report.",
      max_length: 100
    },
    validate: null
  },
  {
    name: "sst",
    description: {
      type: "integer",
      required: false,
      read_only: false,
      label: "Sub-Site Type",
      help_text:
        "Sub-Site Type: describes the primary business function of the sub-site location. This master data attribute is only applicable to a sub-site location.  This value is expressed as a single numerical code.",
      min_value: -32768,
      max_value: 32767
    },
    validate: null
  },
  {
    name: "ssa",
    description: {
      type: "string",
      required: false,
      read_only: false,
      label: "Sub-Site Attribute",
      help_text:
        "Sub-Site Attribute: further qualifies the business function of the sub-site location. This master data attribute is only applicable to a sub-site location. Sub-site attributes are expressed as a comma- separated list of zero or more numerical codes",
      max_length: 1000
    },
    validate: null
  },
  {
    name: "company",
    description: {
      type: "field",
      required: false,
      read_only: false,
      label: "Company",
      help_text: "The company, if any, associated with this location."
    },
    validate: null
  },
  {
    name: "site",
    description: {
      type: "field",
      required: false,
      read_only: false,
      label: "Site",
      help_text:
        "Identifies the site in which this location is contained...if at all. For a Sub-site location, this is the identifier of the parent location."
    },
    validate: null
  },
  {
    name: "location_type",
    description: {
      type: "field",
      required: false,
      read_only: false,
      label: "Location Type",
      help_text:
        "An additional classifier that can be used to identifythe location outside of the CBV codes.  This can be an internal classifier or a human readable that lends further clarity to the location record."
    },
    validate: null
  }
];

// Recharts resizable chart prevents this from working. Skipping for now.
it.skip("renders correctly", () => {
  const form = renderer
    .create(
      <TestWrapper>
        <PageForm
          handleSubmit={() => {
            return null;
          }}
        />
      </TestWrapper>
    )
    .toJSON();
  expect(form).toMatchSnapshot();
});

Enzyme.configure({adapter: new Adapter()});

// Recharts resizable chart prevents this from working. Skipping for now.
it("builds form properly with django options", async () => {
  const promise = Promise.resolve(formattedFields);
  sinon
    .stub(authAPI, "getFormInfo")
    .callsFake((server, path, createForm, processField) => {
      createForm(formattedFields);
      return promise;
    });
  const DecoratedForm = reduxForm({form: "test-form"})(PageForm);
  const form = mount(
    <TestWrapper>
      <DecoratedForm
        handleSubmit={() => {
          return null;
        }}
      />
    </TestWrapper>
  );

  return promise.then(ret => {
    expect(form.find(DecoratedForm).html()).toMatchSnapshot();
  });
});
