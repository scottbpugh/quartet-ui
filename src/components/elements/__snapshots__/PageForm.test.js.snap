// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds form properly with django options 1`] = `"<div><form><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">GLN13</label><div class=\\"pt-form-content\\"><input name=\\"GLN13\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">The GLN (Global Location Number) provides a standard means to identify legal entities, trading parties and locations to support the requirements of electronic commerce. The GLN-13 is defined by GS1</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">SGLN</label><div class=\\"pt-form-content\\"><input name=\\"SGLN\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">The SGLN EPC scheme is used to assign a unique identity to a physical location or sub-location, such as a specific building or a specific unit of shelving within a warehouse.  TheSGLN is expressed as a URN value.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Name</label><div class=\\"pt-form-content\\"><input name=\\"name\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" required=\\"\\" value=\\"\\"><div class=\\"pt-form-helper-text\\">A unique name for the location or party.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Street Address One</label><div class=\\"pt-form-content\\"><input name=\\"address1\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">For example, the name of the street and the number in the street or the name of a building</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Street Address Two</label><div class=\\"pt-form-content\\"><input name=\\"address2\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">The second free form line complements the first free form line to locate the party or location.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Street Address Three</label><div class=\\"pt-form-content\\"><input name=\\"address3\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">The third free form line complements the first and second free form lines where necessary.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Country Code</label><div class=\\"pt-form-content\\"><input name=\\"country\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">Country ISO 3166-1 alpha-2 Code</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">City</label><div class=\\"pt-form-content\\"><input name=\\"city\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">City</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">State or Province</label><div class=\\"pt-form-content\\"><input name=\\"state_province\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">One of the constituent units of a nation having a federal government.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Postal Code</label><div class=\\"pt-form-content\\"><input name=\\"postal_code\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">Postal Code</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Latitude</label><div class=\\"pt-form-content\\"><input name=\\"latitude\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\"> Latitude of the location, in degrees. Positive numbers are northern latitude; negative numbers are southern latitude.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Longitude</label><div class=\\"pt-form-content\\"><input name=\\"longitude\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">Longitude of the location, in degrees. Positive numbers are eastern longitude; negative numbers are western longitude.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Icon</label><div class=\\"pt-form-content\\"><input name=\\"icon\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">An icon to represent the location in a GUI or report.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Sub-Site Type</label><div class=\\"pt-form-content\\"><input name=\\"sst\\" type=\\"number\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">Sub-Site Type: describes the primary business function of the sub-site location. This master data attribute is only applicable to a sub-site location.  This value is expressed as a single numerical code.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Sub-Site Attribute</label><div class=\\"pt-form-content\\"><input name=\\"ssa\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">Sub-Site Attribute: further qualifies the business function of the sub-site location. This master data attribute is only applicable to a sub-site location. Sub-site attributes are expressed as a comma- separated list of zero or more numerical codes</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Company</label><div class=\\"pt-form-content\\"><input name=\\"company\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">The company, if any, associated with this location.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Site</label><div class=\\"pt-form-content\\"><input name=\\"site\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">Identifies the site in which this location is contained...if at all. For a Sub-site location, this is the identifier of the parent location.</div></div></div></div><div style=\\"display: block;\\"><div class=\\"pt-form-group\\"><label class=\\"pt-label\\">Location Type</label><div class=\\"pt-form-content\\"><input name=\\"location_type\\" type=\\"text\\" width=\\"300\\" class=\\"pt-input \\" value=\\"\\"><div class=\\"pt-form-helper-text\\">An additional classifier that can be used to identifythe location outside of the CBV codes.  This can be an internal classifier or a human readable that lends further clarity to the location record.</div></div></div></div><button class=\\"pt-button pt-intent-primary\\" type=\\"submit\\"><span>Submit</span></button> | <button class=\\"pt-button\\" type=\\"button\\"><span>Cancel</span></button></form></div>"`;

exports[`renders correctly 1`] = `
<div>
  <div>
    <form
      onSubmit={null}
    >
      <button
        className="pt-button pt-intent-primary"
        disabled={undefined}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
